crates/nu-cli/src/repl.rs

```diff
@@ -456,18 +456,32 @@ pub fn evaluate_repl(
                         .into_diagnostic()?; // todo: don't stop repl if error here?
                 }
 
-                engine_state
-                    .repl_buffer_state
-                    .lock()
-                    .expect("repl buffer state mutex")
-                    .replace(line_editor.current_buffer_contents().to_string());
-
                 // Right before we start running the code the user gave us,
                 // fire the "pre_execution" hook
                 if let Some(hook) = config.hooks.pre_execution.clone() {
+                    // Set the REPL buffer to the current command for the "pre_execution" hook
+                    let next_repl_buffer = engine_state
+                        .repl_buffer_state
+                        .lock()
+                        .expect("repl buffer state mutex")
```

Comment:
I assume those can only panic if the lock were to be poisoned because a thread holding it was unwinding? Else the worst case is a deadlock.

Thinking about that: The `MutexGuard` is not held across the `eval_hook()` call?

Reply:
I had to read this code a few times and I'm still a little surprised that it doesn't deadlock. 

You're calling `repl_buffer_state.lock()` 3 times in the same scope without manually dropping the `MutexGuard`; I guess it's being implicitly dropped between calls?

Reply:
> I assume those can only panic if the lock were to be poisoned because a thread holding it was unwinding?

Right.

Reply:
> Maybe it would be good to manually `drop()` to make that a little more obvious? This is not an area I have a ton of experience with, I'm open to being corrected.

Good idea. Done now.

Reply:
Thanks! making this more explicit gives some peace of mind. Not everyone has the non-lexical lifetime logic of rust ready at a glance :D